### PR TITLE
Refactor: Refactor LensHub Functionality to Increase Dependency on Libraries for Code Size

### PR DESCRIPTION
**Pull request by Zer0dot**
This PR refactors functionality from the LensHub in favor of libraries. This is a significant change, and includes *cursed* inline assembly to minimize the LensHub's code size.

### CODE REVIEW

File: contracts/core/base/LensHubNFTBase.sol [hunk 1]
```diff
@@ -0,0 +1,40 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.14;
+
+import {ILensNFTBase} from '../../interfaces/ILensNFTBase.sol';
+import {Errors} from '../../libraries/Errors.sol';
```
Comment (miguelmtzinf) [inline contracts/core/base/LensHubNFTBase.sol:6]: Seems not needed

File: contracts/core/base/LensHubNFTBase.sol [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+ * @notice This is an abstract base contract to be inherited by other Lens Protocol NFTs, it includes
+ * the slightly modified ERC721Enumerable, which itself inherits from the ERC721Time-- which adds an
+ * internal operator approval setter, stores the mint timestamp for each token, and replaces the
+ * constructor with an initializer.
+ */
+abstract contract LensHubNFTBase is ERC721Enumerable, ILensNFTBase {
+    mapping(address => uint256) public sigNonces; // Slot 10
+
+    /**
+     * @notice Initializer sets the name, symbol and the cached domain separator.
```
Comment (miguelmtzinf) [inline contracts/core/base/LensHubNFTBase.sol:28]:
```suggestion
     * @notice Initializer sets the name and symbol.
```

File: contracts/core/base/LensMultiState.sol [hunk 3]
```diff
@@ -39,11 +39,11 @@ abstract contract LensMultiState {
...
-    function _setState(DataTypes.ProtocolState newState) internal {
-        DataTypes.ProtocolState prevState = _state;
-        _state = newState;
-        emit Events.StateSet(msg.sender, prevState, newState, block.timestamp);
-    }
+    // function _setState(DataTypes.ProtocolState newState) internal {
+        // DataTypes.ProtocolState prevState = _state;
+        // _state = newState;
+        // emit Events.StateSet(msg.sender, prevState, newState, block.timestamp);
+    // }
```
Comment (miguelmtzinf) [inline contracts/core/base/LensMultiState.sol:46]: Why are you commenting this function out? Is the `StateSet` event still needed?

File: contracts/core/base/LensNFTBase.sol [hunk 4]
```diff
@@ -55,7 +55,7 @@ abstract contract LensNFTBase is ERC721Enumerable, ILensNFTBase {
         address spender,
         uint256 tokenId,
         DataTypes.EIP712Signature calldata sig
-    ) external override {
+    ) external virtual override {
```
Comment (miguelmtzinf) [inline contracts/core/base/LensNFTBase.sol:58]: Seems odd to me there is no `virtual` in `permitForAll()` too

File: contracts/libraries/GeneralLib.sol [hunk 5]
```diff
@@ -0,0 +1,1173 @@
...
+import {Events} from './Events.sol';
+import {IFollowModule} from '../interfaces/IFollowModule.sol';
+import {ICollectModule} from '../interfaces/ICollectModule.sol';
+import {IReferenceModule} from '../interfaces/IReferenceModule.sol';
+
+import './Constants.sol';
+import {MetaTxHelpers} from './MetaTxHelpers.sol';
+import {InteractionHelpers} from './InteractionHelpers.sol';
+
+// TODO: For publishing, increment the pubId here.
```
Comment (miguelmtzinf) [inline contracts/libraries/GeneralLib.sol:18]: ?

Comment (Zer0dot) [reply contracts/libraries/GeneralLib.sol:18]: Removing!

File: contracts/core/base/LensMultiState.sol (hunk 3 above)
Comment (Zer0dot) [reply contracts/core/base/LensMultiState.sol:46]: This logic is delegated to the `GeneralLib` now for code size.

File: contracts/core/base/LensHubNFTBase.sol (hunk 1 above)
Comment (Zer0dot) [reply contracts/core/base/LensHubNFTBase.sol:6]: Removed the Errors import!

File: contracts/core/base/LensHubNFTBase.sol (hunk 2 above)
Comment (Zer0dot) [reply contracts/core/base/LensHubNFTBase.sol]: Fixed, though we should look at caching the domain separator if possible.

**Pull request closed without merging**